lib/payment_processor/clients/braintree/transaction.rb

```diff
@@ -48,12 +48,63 @@ def options
               submit_for_settlement: true,
               store_in_vault_on_success: store_in_vault?
             },
-            customer: {
-              first_name: @user[:first_name] || @user[:name],
-              last_name: @user[:last_name],
-              email: @user[:email]
-            }
-          }.tap{ |opts| opts[:customer_id] = @customer.customer_id if @customer }
+            customer: customer_options,
+            billing: billing_options
+          }.tap do |options|
+            options[:customer_id] = @customer.customer_id if @customer
+          end
+        end
+
+        def customer_options
+          email = @user[:email]
+          email = @customer.email if email.blank? && @customer.present?
+          @customer_options ||= {
+            first_name: @user[:first_name] || namesplitter.first_name,
+            last_name: @user[:last_name] || namesplitter.last_name,
+            email: @user[:email] || email || ''
+          }
+        end
+
+        def billing_options
+          @billing_options ||= {
+            first_name: customer_options[:first_name],
+            last_name: customer_options[:last_name]
+          }.tap do |options|
+            # Yes, this could be done with some nice metaprogramming, but I reckon
+            # this is a but more self-documenting and straightforward
+
+            # `postal_code` can be filled with `postal`, `postal_code`, `zip`, or `zip_code`
+            options[:postal_code]         = @user[:zip] if @user[:zip].present?
+            options[:postal_code]         = @user[:zip_code] if @user[:zip_code].present?
+            options[:postal_code]         = @user[:postal] if @user[:postal].present?
+            options[:postal_code]         = @user[:postal_code] if @user[:postal_code].present?
+
```

Comment:
You must have known I was going to call you on this. No metaprogramming required to make this less of an eye sore.

Just one extra method could make this more expressive:

```ruby

def pick_from(*keys)
  @user.slice(*keys).values.reject(&:blank?).last
end
```
then...
```ruby
options[:postal_code] = pick_from :zip, :zip_code, :postal, :postal_code
```

If you don't want `options` with `nil` values, then call `compact` on it when you're done.